sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -333,31 +335,41 @@ private void takeSample(long currentTimeMillis, long millisSinceLastSample) {
         transitionsAtLastSample = transitionsAtThisSample;
       } else {
         long lullTimeMs = currentTimeMillis - lastTransitionTime.get();
-        Thread thread = trackedThread.get();
         if (lullTimeMs > MAX_LULL_TIME_MS) {
-          if (thread == null) {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for at least %s without outputting or completing (stack trace unable to be generated).",
-                    processBundleId.get(),
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod())));
-          } else if (currentExecutionState == null) {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for at least %s without outputting or completing:%n  at %s",
-                    processBundleId.get(),
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod()),
-                    Joiner.on("\n  at ").join(thread.getStackTrace())));
-          } else {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for PTransform{id=%s, name=%s, state=%s} for at least %s without outputting or completing:%n  at %s",
-                    processBundleId.get(),
-                    currentExecutionState.ptransformId,
-                    currentExecutionState.ptransformUniqueName,
-                    currentExecutionState.stateName,
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod()),
-                    Joiner.on("\n  at ").join(thread.getStackTrace())));
+          if (lullTimeMs < lastLullReport // This must be a new report.
+              || lullTimeMs > 1.2 * lastLullReport // Exponential backoff.
+              || lullTimeMs
+                  > MAX_LULL_TIME_MS + lastLullReport // At least once every MAX_LULL_TIME_MS.
+          ) {
+            lastLullReport = lullTimeMs;
```

Comment:
seems like this should be reset for each processed bundle (ie in start) or should be reset when state changes (ie where transitionsAtLastSample is set) . Otherwise for long-lived streaming pipelines long ops may have happened long before but set a high watermark preventing future logs for things stuck for 5 minutes.

Reply:
The first condition will trigger again.